pinot-core/src/main/java/org/apache/pinot/core/plan/CombinePlanNode.java

```diff
@@ -73,87 +76,33 @@ public CombinePlanNode(List<PlanNode> planNodes, QueryContext queryContext, Exec
-      // returns. We need to ensure no execution left before the main thread returning because the main thread holds the
-      // reference to the segments, and if the segments are deleted/refreshed, the segments can be released after the
-      // main thread returns, which would lead to undefined behavior (even JVM crash) when executing queries against
-      // them.
-      Phaser phaser = new Phaser(1);
-
-      // Submit all jobs
-      Future[] futures = new Future[numTasks];
-      for (int i = 0; i < numTasks; i++) {
-        int index = i;
-        futures[i] = _executorService.submit(new TraceCallable<List<Operator>>() {
-          @Override
-          public List<Operator> callJob() {
-            try {
-              // Register the thread to the phaser.
-              // If the phaser is terminated (returning negative value) when trying to register the thread, that means
-              // the query execution has timed out, and the main thread has deregistered itself and returned the result.
-              // Directly return as no execution result will be taken.
-              if (phaser.register() < 0) {
-                return Collections.emptyList();
-              }
-
-              List<Operator> operators = new ArrayList<>();
-              for (int i = index; i < numPlanNodes; i += numTasks) {
-                operators.add(_planNodes.get(i).run());
-              }
-              return operators;
-            } finally {
-              phaser.arriveAndDeregister();
-            }
-          }
-        });
-      }
-
-      // Get all results
-      try {
-        for (Future future : futures) {
-          List<Operator> ops = (List<Operator>) future.get(_queryContext.getEndTimeMs() - System.currentTimeMillis(),
-              TimeUnit.MILLISECONDS);
-          operators.addAll(ops);
-        }
-      } catch (Exception e) {
-        // Future object will throw ExecutionException for execution exception, need to check the cause to determine
-        // whether it is caused by bad query
-        Throwable cause = e.getCause();
-        if (cause instanceof BadQueryRequestException) {
-          throw (BadQueryRequestException) cause;
-        } else {
-          throw new RuntimeException("Caught exception while running CombinePlanNode.", e);
-        }
-      } finally {
-        // Cancel all ongoing jobs
-        for (Future future : futures) {
-          if (!future.isDone()) {
-            future.cancel(true);
-          }
-        }
-        // Deregister the main thread and wait for all threads done
-        phaser.awaitAdvance(phaser.arriveAndDeregister());
+      if (maxExecutionThreads == 1) {
+        operators = getOperatorsSingleThreaded();
+      } else {
+        int numTasks = Math.min((numPlanNodes + TARGET_NUM_PLANS_PER_THREAD - 1) / TARGET_NUM_PLANS_PER_THREAD,
+            maxExecutionThreads);
+        scope.setNumTasks(numTasks);
```

Comment:
Shall we also record numOperators (which should be equal to the number of segments) ?

Reply:
That's already recorded on line 88